### PR TITLE
[OpenMP] Add num_threads clause list format and strict modifier support

### PR DESCRIPTION
**Pull request by TerryLWilmarth**
Add support to the runtime for 6.0 spec features that allow num_threads clause to take a list, and also make use of the strict modifier.  Provides new compiler interface functions for these features.

### CODE REVIEW

**Comment by shiltian** [comment]
Can we have tests?

File: openmp/runtime/src/dllexports [hunk 1]
```diff
@@ -1268,6 +1268,11 @@ kmp_set_disp_num_buffers                    890
     __kmpc_atomic_val_8_cas_cpt            2158
     %endif
 
+    # No longer need to put ordinal numbers
```
Comment (shiltian) [inline openmp/runtime/src/dllexports:1271]: This will not cause issue on Windows anymore?

**Review by shiltian — APPROVED** [review]
LGTM. Will the front end support be upstreamed as well?

File: openmp/runtime/src/dllexports (hunk 1 above)
Comment (jpeyton52) [reply openmp/runtime/src/dllexports:1271]: Right, you can add exported symbols just by name now. Older symbols keep their ordinal number.

**Review by jpeyton52 — APPROVED** [review]
LGTM

**Comment by mstorsjo** [comment]
This change causes failed tests on i386: https://github.com/mstorsjo/llvm-mingw/actions/runs/9654575386/job/26643787120

```
# .---command stderr------------
# | OMP: Error #137: Cannot create thread.
# | OMP: System error #8: (No system error message available)
# `-----------------------------
# error: command failed with exit status: 3
```

I presume the test spawns a very large number of threads? Can it be marked with `UNSUPPORTED` for 32 bit environments somehow?

**Comment by TerryLWilmarth** [comment]
> This change causes failed tests on i386: https://github.com/mstorsjo/llvm-mingw/actions/runs/9654575386/job/26643787120
> 
> ```
> # .---command stderr------------
> # | OMP: Error #137: Cannot create thread.
> # | OMP: System error #8: (No system error message available)
> # `-----------------------------
> # error: command failed with exit status: 3
> ```
> 
> I presume the test spawns a very large number of threads? Can it be marked with `UNSUPPORTED` for 32 bit environments somehow?

Yes, in the second to last case, the test uses a pretty deep level of nesting, and tests the num_threads clause list at two different nesting levels, so it will generate a lot of threads.  We could modify the test so that the list used at the outer level is something like 3,1 and inner level is 3,1, which should create 9 threads at most.  That should be adequate to show the feature is working correctly.  Would that work for this system?

**Comment by mstorsjo** [comment]
> > This change causes failed tests on i386: https://github.com/mstorsjo/llvm-mingw/actions/runs/9654575386/job/26643787120
> > ```
> > # .---command stderr------------
> > # | OMP: Error #137: Cannot create thread.
> > # | OMP: System error #8: (No system error message available)
> > # `-----------------------------
> > # error: command failed with exit status: 3
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I presume the test spawns a very large number of threads? Can it be marked with `UNSUPPORTED` for 32 bit environments somehow?
> 
> Yes, in the second to last case, the test uses a pretty deep level of nesting, and tests the num_threads clause list at two different nesting levels, so it will generate a lot of threads. We could modify the test so that the list used at the outer level is something like 3,1 and inner level is 3,1, which should create 9 threads at most. That should be adequate to show the feature is working correctly. Would that work for this system?

I'm pretty sure that 9 threads should be quite doable to spawn. If you put up a PR to do this modification, I might be able to test it.

**Comment by mstorsjo** [comment]
Ping - can you post the change (or provide a link to it?) for reducing the number of threads here, to get the tests back to passing again?

**Comment by TerryLWilmarth** [comment]
Working on it.

**Comment by TerryLWilmarth** [comment]
Martin, will it work with 36 threads? I have a pull request for that change up now.  Using 1 in the outer level list has exposed what appears to be a non-trivial bug. I'll work on that and add the triggering test case when fixed.  Let me know if the 36 threads case will work for you for now. Thanks!

**Pull request merged**